src/core/symbology-ng/qgsfillsymbollayerv2.cpp

```diff
@@ -3430,6 +3436,40 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
 {
   Q_UNUSED( rings );
 
+  if ( !mPointOnAllParts )
+  {
+    const QgsFeature* feature = context.feature();
+    if ( feature )
+    {
+      if ( feature->id() != mBiggestPartFeatureId )
+      {
+        mBiggestPartFeatureId = feature->id();
+        mBiggestPartIndex = 1;
+
+        if ( context.mGeometryPartCount > 1 )
+        {
+          const QgsGeometry *geom = feature->constGeometry();
+          const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+          double area;
+          double area_biggest = 0;
+          for ( int i = 0; i < context.mGeometryPartCount; ++i )
+          {
+            area = geomCollection->geometryN( i )->area();
+            if ( area > area_biggest )
+            {
+              area_biggest = area;
+              mBiggestPartIndex = i + 1;
+            }
+          }
+        }
+      }
+
+      if ( context.mGeometryPartNum != mBiggestPartIndex )
```

Comment:
Could this be moved 3 lines down in the form

    if ( mPointOnAllParts || context.mGeometryPartNum == mBiggestPartIndex )
    {
      QPointF centroid = mPointOnSurface ? QgsSymbolLayerV2Utils::polygonPointOnSurface( points ) : QgsSymbolLayerV2Utils::polygonCentroid( points );
      mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );		    mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );
    }